src/chainstate/coordinator/mod.rs

```diff
@@ -0,0 +1,363 @@
+            //  about to process is the first block in reward cycle.
+            let reward_cycle_info = self.get_reward_cycle_info(&header);
+            let sortition_id = self.sortition_db.evaluate_sortition(
+                &header, ops, &self.burnchain, &canonical_sortition_tip, reward_cycle_info)
+                .map_err(|e| {
+                    error!("ChainsCoordinator: unable to evaluate sortition {:?}", e);
+                    Error::FailedToProcessSortition(e)
+                })?
+                .0.sortition_id;
+
+            self.notifier.notify_sortition_processed();
+
+            debug!("Sortition processed: {}", &sortition_id);
+
+            if sortition_tip_snapshot.block_height < header.block_height {
+                // bump canonical sortition...
+                self.canonical_sortition_tip = Some(sortition_id.clone());
+                canonical_sortition_tip = sortition_id;
+            }
+
+            if let Some(pox_anchor) = self.process_ready_blocks()? {
+                return self.process_new_pox_anchor(&pox_anchor)
+            }
+        }
+
+        Ok(())
+    }
+
+    /// returns None if this burnchain block is _not_ the start of a reward cycle
+    ///         otherwise, returns the required reward cycle info for this burnchain block
+    ///                     in our current sortition view:
+    ///           * PoX anchor block
+    ///           * Was PoX anchor block known?
+    fn get_reward_cycle_info(&self, burn_header: &BurnchainBlockHeader) -> Option<RewardCycleInfo> {
+        if self.burnchain.is_reward_cycle_start(burn_header.block_height) {
+            info!("Beginning reward cycle. block_height={}", burn_header.block_height);
+            Some(RewardCycleInfo {
+                anchor_block: None,
+                anchor_block_known: true
+            })
+        } else {
+            None
+        }
+    }
+
+    ///
+    /// Process any ready staging blocks until there are no more to process
+    ///  _or_ a PoX anchor block is discovered.
+    ///
+    /// Returns Some(StacksBlockId) if an anchor block is discovered,
+    ///   otherwise returns None
+    fn process_ready_blocks(&mut self) -> Result<Option<BlockHeaderHash>, Error> {
+        let canonical_sortition_tip = self.canonical_sortition_tip.as_ref()
+            .expect("FAIL: processing a new Stacks block, but don't have a canonical sortition tip");
+
+        let sortdb_handle = self.sortition_db.tx_handle_begin(canonical_sortition_tip)?;
+        let mut processed_blocks = self.chain_state_db.process_blocks(sortdb_handle, 1)?;
+
+        while let Some(block_result) = processed_blocks.pop() {
+            if let (Some(block_receipt), _) = block_result {
+                // only bump the coordinator's state if the processed block
+                //   is in our sortition fork
+                //  TODO: we should update the staging block logic to prevent
```

Comment:
The staging block DB primarily cares about block storage and retrieval.  It should use a handle to the sortition DB only as a way to get hints on whether or not to store or evict a block based on whether it is present in the PoX fork.  I'd rather have the PoX state kept in one place instead of try and share it across the sortition DB and the staging blocks DB.

Reply:
Agreed, and that's more or less what this comment is suggesting. We don't want the staging block DB to accept Stacks blocks which aren't in our PoX fork, but we should also at some point be proactive about evicting previously accepted blocks when a PoX fork becomes invalid.

Reply:
So is this TODO more about the need to evict now-invalidated blocks?  Because I think it already prevents blocks that aren't on the PoX canonical fork from being stored; it just doesn't get rid of them later if the PoX fork gets invalidated.

Reply:
Yep

Reply:
Cool. Can you open an issue for this and resolve this conversation when you have done so?

Reply:
https://github.com/blockstack/stacks-blockchain/issues/1804